app/workers/calculate_project_completeness_worker.rb

```diff
@@ -50,7 +57,7 @@ def workflow_completeness(workflow)
 
     retired_subjects = workflow.retired_subjects_count
     total_subjects = workflow_subjects_count
-    (0.0..1.0).clamp(retired_subjects / total_subjects.to_f)
+    (0.0..1.0).clamp(retired_subjects / total_subjects.to_d)
```

Comment:
Wanted to double check this, is there ever a case `total_subjects` is ever not a whole number? Are we using `to_d` in this case for precision of the calculation `retired_subjects / total_subjects`?

Reply:
yes - I switched to .to_d `BigDecimal` class to avoid the float precision issues in the resulting calculations and kept it consistent in the file (i.e. avoid mixing .to_f and .to_d)

Details outlined here https://www.rubyguides.com/2016/07/numbers-in-ruby/  in the `Float Imprecision` section.

Reply:
> Wanted to double check this, is there ever a case total_subjects is ever not a whole number? 

Workflow.`subjects_count` wraps the db attribute `real_set_member_subjects_count` at https://github.com/zooniverse/panoptes/blob/2fea4d095ca46436d8faac466baa8f732e7bd531/app/models/workflow.rb#L138 and that has a not null default 0 set on the db attribute, https://github.com/zooniverse/panoptes/blob/2fea4d095ca46436d8faac466baa8f732e7bd531/db/structure.sql#L1964

So that value should always be a whole number that is set by the workflow counters as per https://github.com/zooniverse/panoptes/blob/2fea4d095ca46436d8faac466baa8f732e7bd531/app/counters/workflow_counter.rb#L17